lego/apps/polls/serializers.py

```diff
@@ -0,0 +1,151 @@
+
+    has_answered = serializers.SerializerMethodField()
+
+    def get_has_answered(self, obj):
+        user = self.context["request"].user
+        return obj.get_has_answered(user)
+
+    class Meta:
+        model = Poll
+        fields = (
+            "id",
+            "created_at",
+            "valid_until",
+            "title",
+            "description",
+            "options",
+            "total_votes",
+            "tags",
+            "comments",
+            "comment_target",
+            "has_answered",
+            "pinned",
+        )
+
+
+class DetailedPollSerializer(TagSerializerMixin, BasisModelSerializer):
+
+    comments = CommentSerializer(read_only=True, many=True)
+    comment_target = CharField(read_only=True)
+
+    options = OptionSerializer(many=True)
+    total_votes = IntegerField(read_only=True)
+
+    has_answered = serializers.SerializerMethodField()
+
+    def get_has_answered(self, obj):
+        user = self.context["request"].user
+        return obj.get_has_answered(user)
+
+    class Meta:
+        model = Poll
+        fields = (
+            "id",
+            "created_at",
+            "valid_until",
+            "title",
+            "description",
+            "options",
+            "total_votes",
+            "comments",
+            "comment_target",
+            "tags",
+            "has_answered",
+            "pinned",
+        )
+
+
+class PollCreateSerializer(TagSerializerMixin, BasisModelSerializer):
+
+    options = OptionSerializer(many=True)
+
+    class Meta:
+        model = Poll
+        fields = (
+            "id",
+            "created_at",
+            "title",
+            "description",
+            "options",
+            "total_votes",
+            "tags",
+            "pinned",
+            "valid_until",
+        )
+
+    @transaction.atomic
+    def create(self, validated_data):
+        options_data = validated_data.pop("options")
+        poll = super().create(validated_data)
+        for option in options_data:
+            Option.objects.create(poll=poll, **option)
+        return poll
+
+
+class PollUpdateSerializer(TagSerializerMixin, BasisModelSerializer):
+
+    options = OptionUpdateSerializer(many=True)
+
+    class Meta:
+        model = Poll
+        fields = (
+            "id",
+            "created_at",
+            "title",
+            "description",
+            "options",
+            "total_votes",
+            "tags",
+            "pinned",
+        )
+
+    @transaction.atomic
+    def update(self, instance, validated_data):
+        options = validated_data.pop("options")
+
+        # Update the regular fields that aren't options first
+        super().update(instance, validated_data)
+
+        # Delete options that aren't in the received list
+        for old_option in instance.options.all():
+            options_with_ids = filter(lambda q: "id" in q, options)
+            existing_options_ids = map(lambda q: q["id"], options_with_ids)
```

Comment:
Could we move these two lines to before the loop as they don't need to be run each iteration?

Reply:
yup